okhttp/src/commonMain/kotlin/okhttp3/internal/-HostnamesCommon.kt

```diff
@@ -225,4 +225,89 @@ internal fun inet6AddressToAscii(address: ByteArray): String {
   return result.readUtf8()
 }
 
-internal expect fun String.toCanonicalHost(): String?
+/**
+ * Returns the canonical address for [address]. If [address] is an IPv6 address that is mapped to an
+ * IPv4 address, this returns the IPv4-mapped address. Otherwise, this returns [address].
+ *
+ * https://en.wikipedia.org/wiki/IPv6#IPv4-mapped_IPv6_addresses
+ */
+internal fun canonicalizeInetAddress(address: ByteArray): ByteArray {
+  return when {
+    isMappedIpv4Address(address) -> address.sliceArray(12 until 16)
+    else -> address
+  }
+}
+
+/** Returns true for IPv6 addresses like `0000:0000:0000:0000:0000:ffff:XXXX:XXXX`. */
+private fun isMappedIpv4Address(address: ByteArray): Boolean {
```

Comment:
New code! So I don’t use `InetAddress.getByAddress()` to perform IPv6 to IPv4 mapping.